fuel-vm/src/interpreter/executors/instruction.rs

```diff
@@ -50,446 +50,539 @@ where
             }
         }
 
-        self._instruction(instruction)
-            .map_err(|e| InterpreterError::from_runtime(e, instruction))
+        self._instruction(raw)
+            .map_err(|e| InterpreterError::from_runtime(e, raw))
     }
 
     #[tracing::instrument(name = "instruction", skip(self))]
-    fn _instruction(&mut self, instruction: Instruction) -> Result<ExecuteState, RuntimeError> {
-        let (op, ra, rb, rc, rd, imm) = instruction.into_inner();
-        let (a, b, c, d) = (
-            self.registers[ra],
-            self.registers[rb],
-            self.registers[rc],
-            self.registers[rd],
-        );
+    fn _instruction(&mut self, raw: RawInstruction) -> Result<ExecuteState, RuntimeError> {
+        let instruction = Instruction::try_from(raw).map_err(|_| RuntimeError::from(PanicReason::ErrorFlag))?;
 
-        tracing::trace!("Op code: {:?}, Registers: a {}, b, {}, c {}, d {}", op, a, b, c, d);
+        tracing::trace!("Instruction: {:?}", instruction);
 
         // TODO additional branch that might be optimized after
         // https://github.com/FuelLabs/fuel-asm/issues/68
-        if self.is_predicate() && !op.is_predicate_allowed() {
+        if self.is_predicate() && !instruction.opcode().is_predicate_allowed() {
             return Err(PanicReason::TransactionValidity.into());
         }
 
-        match op {
-            OpcodeRepr::ADD => {
+        // Short-hand for retrieving the value from the register with the given ID.
+        // We use a macro to "close over" `self.registers` without taking ownership of it.
+        macro_rules! r {
```

Comment:
is a macro really necessary for doing this basic fetch? Couldn't we achieve similar results with an inlined helper function that takes `&self` as the first arg?

Reply:
It's repeated so much in the rest of the function that I feel usage of a simple macro here more readable. We would have to replace `r!(a).into()` with `self.reg(a).into()` which isn't any more readable for me.